### ISSUE TITLE
Using types that are exported indirectly via `typeof`

### ISSUE THREAD

**Issue by fabiospampinato** [Microsoft/TypeScript#28335]
I'm importing a file that has the following types declaration:

```typescript
declare class FIFO {}
declare const _default: typeof FIFO & {
    default: typeof FIFO;
};
export = _default;
```

And I want to use it like this:

```typescript
import FIFO from '/path/to/fifo';
export default class Foo {
  queue = new FIFO ()
};
```

I don't think this code has any problem whatsoever, i.e. I would be able to generate proper types for this file, but TypeScript v3.1.6 is complaining with:

```
Public property 'queue' of exported class has or is using name 'FIFO' from external module "/path/to/fifo" but cannot be named.
```

What am I supposed to do? I can't export the `fifo` module in a different manner, and I want proper types declarations generated for my file.

**Comment by fabiospampinato** [comment]
@weswigham why is this a necessity? I, as a human, can see that the default export if of `typeof FIFO`, and a few lines above in the same file I can see what `FIFO`'s actual type is. 

Why can't TypeScript be smarter and do the same?

Also I can't export `FIFO` in a different way, is TypeScript just not supposed to work with this perfectly fine code that exports type declarations?

**Comment by weswigham** [comment]
@fabiospampinato only types which are explicitly `export`ed can be seen by other files. This is done for the same reason that, for example, at runtime you can only get the exports of another module and not its locals. Privates are used for representing internal structures.

**Comment by fabiospampinato** [comment]
> This is done for the same reason that, for example, at runtime you can only get the exports of another module and not its locals. Privates are used for representing internal structures.

I'm not sold on this analogy, if I have an internal non-primitive variable called `myVar` which I export with `export = { default: myVar }` I can access `myVar` from the outside, so I don't think this counts as "internal". Similarly in my example there is an internal type called `FIFO`, and it's being exported as `export = { default: typeof FIFO }` basically, maybe it's not the same exact thing as doing `export = FIFO`, but that type is not internal in the sense that's inaccessible from the outside, otherwise my editor wouldn't be able to provide me proper autocompletion, right? I don't understand what's the difference here.

You mentioned I have to add something like this to the types:

```typescript
declare namespace _default {
  export type FIFOInstance = FIFO;
}
```

Currently that `FIFO` module is exported with `export = Object.assign ( FIFO, { default: FIFO } )`, how should I change this line so that proper types are generated and backwards compatibility is preserved?

And why do I have to rename the type to `FIFOInstance`?

Thank you!

**Comment by weswigham** [comment]
>export = { default: typeof FIFO } 

Only exports the static part of the class - the shape of it's constructor. It's instance type exists insofar as it's the result of `new`'ing that, but it lacks an exported, accessible name through which it can be referred to in a declaration file. 

> And why do I have to rename the type to FIFOInstance?

`export type FIFO = FIFO` won't work because the `export` declaration of the same name will shadow the access. You can also alias it outside the namespace, then refer to the alias within the namespace to export it with the "correct" name.

>Currently that FIFO module is exported with export = Object.assign ( FIFO, { default: FIFO } ), how should I change this line so that proper types are generated and backwards compatibility is preserved?

Along the lines of
```ts
const exportedFIFO = Object.assign ( FIFO, { default: FIFO } );
type FIFOInstance = FIFO;
namespace exportedFIFO {
  export type FIFO = FIFOInstance;
}
export = exportedFIFO;
```

 Should work.

**Comment by fabiospampinato** [comment]
> Only exports the static part of the class - the shape of it's constructor. It's instance type exists insofar as it's the result of new'ing that, but it lacks an exported, accessible name through which it can be referred to in a declaration file.

That made it clearer for me, thanks, but I still don't get why I can't reference to the type of its instance, I get that it's not _explicitly_ exported, but I'm exporting the constructor, and we know what the type of its instances would be, why do I have to _explicitly_ export it? Shouldn't this be inferred? If there's any problem that may arise from inferring it I'm not able to find one.

> Along the lines of (...) Should work.

Thanks for the suggestion, but going from a clean 1 line export to that seems excessive to me 😩 I just want to export my class in a way that's importable as `FIFO = require ( 'fifo' )` from node and as `import FIFO from 'fifo'` from typescript, basically a class with a static property named `default` that references itself, which is totally valid code, it's a bit frustrating that I can't do this cleanly while still having proper types generated.

**Comment by weswigham** [comment]
> Thanks for the suggestion, but going from a clean 1 line export to that seems excessive to me 😩 I just want to export my class in a way that's importable as FIFO = require ( 'fifo' ) from node and as import FIFO from 'fifo' from typescript, basically a class with a static property named default that references itself, which is totally valid code, it's a bit frustrating that I can't do this cleanly while still having proper types generated

You should just be writing `export = FIFO` and using the `esModuleInterop` flag then, no?

**Comment by fabiospampinato** [comment]
@weswigham That's not backwards compatible, I can't just require everyone involved with this to add `"esModuleInterop": true` to their `tsconfig.json`.

**Comment by weswigham** [comment]
In that case, you'll definitely need to do the long-winded thing. While `FIFO` is a class, it has both a type and a value and an export assignment exports both; however once you pass it thru an `Object.assign`, that expression (as a whole) just has a value - the interface associated with the class instance is still only stored within the reference to the class itself. You'll have to patch that back into your exported namespace at a name you deem appropriate by explicitly reexporting it as I demonstrated above.

**Comment by fabiospampinato** [comment]
It seems there's no cleaner solution via TypeScript alone then, perhaps a babel transform or something could give me what I want without requiring super ugly exports.

Thanks for the help!

**Comment by fabiospampinato** [comment]
In case somebody else stumbles upon this, I've made [typescript-transform-export-interop](https://github.com/fabiospampinato/typescript-transform-export-interop) to greatly alleviate the problem:

1. Just export your modules as `export default Foo` (if you need a default export)
2. And add this module to your compilation chain: `tsc && tstei`, that's it.

@weswigham I'm obviously no expert on this, but it really doesn't make sense to me that types are properly exported if I write `export {Foo}` but not if I write `export = {Foo}`. One shouldn't have to write ugly code or use external tools just to have its modules easily importable from TS & Node.
